[Merged by Bors] - feat(gcongr): fix pattern elaborator and @[gcongr] for lemmas introducing binders

JovanGerb opened the pull request:
This PR adds @[gcongr] tags so that
- `gcongr` can move inside Eventually and Frequently,
- moving inside of `∀ ε > 0,` or `∃ ε > 0,` adds the hypothesis/condition to the local context.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg reviewed (commented):
Thanks! Are there some tests you can add easily?

This seems reasonable, but a second opinion would be good.
maintainer merge?

JovanGerb commented:
Yes, I've added some test to the `peel` tactic file. This is the first step towards replacing `peel` with a tactic built on `gcongr`.

Additionally, in writing the tests, I discovered that `gcongr` uses the wrong elaborator for its pattern: I've replaced `Term.elabTerm` with `Term.elabPattern`. The "large-import" is importing one file that contains only the definion of `elabPattern`.

grunweg commented:
Thanks for the added tests! Does the elaborator change need separate tests, or is it covered by the new ones?

The elaborator change is above my current knowledge (so will need somebody else to review). Everything else looks great to me. Thanks for working on this!
maintainer merge?

JovanGerb commented:
I've added a test for the new pattern elaborator

fpvandoorn commented:
LGTM

Please merge master.

If you can do this now without any additional imports anywhere, please merge. Otherwise, please allow me to have another look.

bors d+

JovanGerb commented:
bors r+

I'm having some trouble merging master due to the github changes, but I don't think merging master would fail here.
